src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSWebWorker.cs

```diff
@@ -30,220 +30,205 @@ public static Task RunAsync(Func<Task> body)
+                else
+                {
+                    // if cancelation was requested, the JSSynchronizationContext will stop the thread
+                    if (_jsSynchronizationContext != null && !_cancellationToken.IsCancellationRequested)
+                    {
+                        // we can only uninstall JS interop on it's own thread
+                        _jsSynchronizationContext.Post(static o =>
+                        {
+                            var jsSynchronizationContext = (JSSynchronizationContext)o!;
+                            // this will lead to __emscripten_thread_exit() later
+                            jsSynchronizationContext?.UninstallWebWorkerInterop();
+                        }, _jsSynchronizationContext);
+                    }
+                    _jsSynchronizationContext = null;
 
-        private static void SendWhenException<T>(SynchronizationContext ctx, TaskCompletionSource<T> tcs, Exception ex)
-        {
-            try
-            {
-                ctx.Send((_) => tcs.SetException(ex), null);
-            }
-            catch (Exception e)
-            {
-                Environment.FailFast("JSWebWorker.RunAsync failed", e);
-            }
-        }
+                    PropagateCompletion();
+                }
 
-        private static void SendWhenDone<T>(SynchronizationContext ctx, TaskCompletionSource<T> tcs, Task<T> done)
-        {
-            try
-            {
-                ctx.Send((_) =>
-                {
-                    PropagateCompletion(tcs, done);
-                }, null);
+                Dispose();
             }
-            catch (Exception e)
-            {
-                Environment.FailFast("JSWebWorker.RunAsync failed", e);
-            }
-        }
 
-        internal static void PropagateCompletion<T>(TaskCompletionSource<T> tcs, Task<T> done)
-        {
-            if (done.IsFaulted)
+            private void PropagateCompletion()
             {
-                if (done.Exception is AggregateException ag && ag.InnerException != null)
+                if (_result!.IsFaulted)
+                {
+                    if (_result.Exception is AggregateException ag && ag.InnerException != null)
+                    {
+                        _taskCompletionSource.TrySetException(ag.InnerException);
+                    }
+                    else
+                    {
+                        _taskCompletionSource.TrySetException(_result.Exception);
+                    }
+                }
+                else if (_result.IsCanceled)
                 {
-                    tcs.SetException(ag.InnerException);
+                    _taskCompletionSource.TrySetCanceled();
                 }
                 else
                 {
-                    tcs.SetException(done.Exception);
+                    if (_bodyRes != null)
+                    {
+                        _taskCompletionSource.TrySetResult(((Task<T>)_result).Result);
```

Comment:
is it possible to get here when _result doesn't have a result? it looks like it could be, but i'm also not super worried about it

Reply:
this is `Task` or `Task<T>` not the resolved value. I will rename it.